File: updater/src/main/java/de/darkatra/patcher/updater/service/model/ObsoleteFile.java
package de.darkatra.patcher.updater.service.model;

import lombok.Data;
import lombok.experimental.Accessors;

import java.nio.file.Paths;

@Data
@Accessors(chain = true)
public class ObsoleteFile implements ContextAware {
	private String dest;

	@Override
	public void applyContext(final Context context) {
		context.forEach((key, value) -> {
			dest = dest.replace(context.getPrefix() + key + context.getSuffix(), value);
		});
		dest = Paths.get(dest).normalize().toString();
	}
}

File: updater/src/main/kotlin/de/darkatra/bfme2/patch/ObsoleteFile.kt
package de.darkatra.bfme2.patch

import io.goodforgod.graalvm.hint.annotation.ReflectionHint
import java.nio.file.Path

@ReflectionHint(ReflectionHint.AccessType.ALL_DECLARED_CONSTRUCTORS)
data class ObsoleteFile(
    var dest: String
) : ContextAware {

    /**
     * Applies the given [Context] to the [ObsoleteFile], replacing placeholders with their actual value.
     * Ideally this operation should only be performed once per [ObsoleteFile].
     *
     * @param context the [Context] containing values for placeholders
     */
    override fun applyContext(context: Context) {

        val origDest: String = dest

        context.forEach { (key, value) ->
            dest = Path.of(dest.replace("${Context.PREFIX}$key${Context.SUFFIX}", value)).normalize().toString()
        }

        if (dest == origDest) {
            error("Invalid ObsoleteFile without placeholder was found. (Dest: '$dest')")
        }
    }
}
